mogp_gpu/src/gp_gpu.hpp

```diff
@@ -0,0 +1,764 @@
+            std::string msg;
+            std::stringstream smsg(msg);
+            smsg << "Error in potrf: return code " << status << ", info " << info_h;
+            throw std::runtime_error(smsg.str());
+
+	}
+	return info_h;
+    }
+
+    // return the lower triangular matrix (actually it will be the whole matrix
+    // but only the lower triangle will be correct!)
+    void get_cholesky_lower(mat_ref result)
+    {
+        thrust::copy(chol_lower_d.begin(), chol_lower_d.end(), result.data());
+    }
+
+    // Update the hyperparameters, and invQ and invQt which depend on them
+    void fit(vec_ref theta, nugget_type nugget, double nugget_size=0.0)
+    {
+        thrust::copy(theta.data(), theta.data() + D + 2, theta_d.begin());
+
+        cov_batch_gpu(dev_ptr(invQ_d), n, n, D, dev_ptr(inputs_d),
+                      dev_ptr(inputs_d), dev_ptr(theta_d));
+
+	/// copy the covariance matrix invQ_d into work_mat_d
+	thrust::copy(invQ_d.begin(), invQ_d.end(), work_mat_d.begin());
+
+        thrust::device_vector<int> info_d(1);
+        int info_h;
+        cusolverStatus_t status;
+	int factorisation_status;
+
+	// for adaptive nugget start with a nugget of zero and increase by small amount
+	// until we find a value where factorization succeeds.
+	if (nugget == NUG_ADAPTIVE) {
+	    double mean_diag;
+	    const int max_tries = 5;
+	    int itry = 0;
+	    while (itry < max_tries) {
+		// invQ_d holds the covariance matrix - work with a copy
+		// in work_mat_d, in case the factorization fails
+		thrust::copy(invQ_d.begin(), invQ_d.end(), work_mat_d.begin());
+
+		// if the first attempt at factorization failed, add a
+		// small term to the diagonal, increasing each iteration
+		// until the factorization succeeds
+		if (itry >= 1) {
+		    if (itry == 1) {
+			// find mean of (absolute) diagonal elements (diagonal
+			// elements should all be positive)
+			cublasDasum(cublasHandle, n, dev_ptr(invQ_d), n+1, &mean_diag);
+			mean_diag /= n;
+			jitter = 1e-6 * mean_diag;
+		    }
+		    add_diagonal(n, jitter, dev_ptr(work_mat_d));
+		}
+
+		factorisation_status = calc_cholesky_factors();
+		if (factorisation_status == 0) {
+                    break;
+		}
+		jitter *= 10;
+		itry++;
+	    }
+
+	    // if none of the factorization attempts succeeded:
+	    if (itry == max_tries) {
+		std::string msg;
+		std::stringstream smsg(msg);
+		smsg << "All attempts at factorization failed. Last return code " << factorisation_status;
+		throw std::runtime_error(smsg.str());
+	    }
+	// for fixed nugget, add "nugget_size" to the diagonal of the matrix.
+	} else if (nugget == NUG_FIXED) {
+	    add_diagonal(n, nugget_size, dev_ptr(work_mat_d));
+
+	    factorisation_status = calc_cholesky_factors();
+	    if (factorisation_status != 0) {
+                throw std::runtime_error("Unable to factorize matrix using fixed nugget");
+	    }
+
+	} else { //nugget == "fit"
+	    factorisation_status = calc_cholesky_factors();
```

Comment:
I'm confused by the fitted nugget code. It doesn't look like the nugget is added here, and I didn't think I saw it added when computing the covariance matrix. Am I missing something?

Reply:
No, you're right - I hadn't implemented the fitted nugget.   Working on this now.

Reply:
Fitted nugget is now implemented 🎉   
Added a new method in `gp_gpu.hpp` to calculate the trace (following approach in `sum_log_diag`) and use this in `logpost_deriv`.
Verified that we get identical logpost_deriv values as CPU version, and also that `fit_GP_MAP` gives identical theta values as CPU version, when using fitted nugget.